File: platforms/linkedin.py
"""LinkedIn — persistent session login, job search, Easy Apply automation."""
import os
import time
import random

try:
    from selenium.webdriver.common.by import By
    from selenium.webdriver.support.ui import WebDriverWait, Select
    from selenium.webdriver.support import expected_conditions as EC
    from selenium.common.exceptions import (
        TimeoutException, NoSuchElementException,
        ElementClickInterceptedException, WebDriverException,
        StaleElementReferenceException,
    )
    _SELENIUM_OK = True
except ImportError:
    _SELENIUM_OK = False  # Vercel — browser methods won't be called

from core.browser import get_driver, human_delay, safe_click, human_type
from core.config import Config
from core.logger import get_logger
from platforms.base import BasePlatform, JobListing

log = get_logger("linkedin")
BASE = "https://www.linkedin.com"


class LinkedInPlatform(BasePlatform):
    name = "linkedin"

    def __init__(self):
        self.driver = None

    def _init(self):
        if not self.driver:
            self.driver = get_driver("linkedin")

    def _is_alive(self) -> bool:
        try:
            _ = self.driver.current_url
            return True
        except Exception:
            return False

    # ── Authentication ──────────────────────────────────────────
    def login(self) -> bool:
        self._init()
        try:
            self.driver.get(f"{BASE}/feed/")
            human_delay(3, 5)
        except Exception:
            return False

        if self._logged_in():
            log.info("LinkedIn: already authenticated")
            return True

        # CI/GitHub Actions: use password login if env var is set
        password = os.getenv("LINKEDIN_PASSWORD", "")
        if password:
            return self._login_with_password(password)

        log.info("LinkedIn: attempting login via saved session")
        self.driver.get(f"{BASE}/login")
        human_delay(2, 3)

        try:
            email_field = WebDriverWait(self.driver, 8).until(
                EC.presence_of_element_located((By.ID, "username"))
            )
            human_type(email_field, Config.CANDIDATE_EMAIL)
            human_delay(0.5, 1.2)
        except Exception:
            pass

        for _ in range(45):
            human_delay(2, 2)
            if self._logged_in():
                log.info("LinkedIn: login successful")
                return True

        log.warning("LinkedIn: not logged in — run setup_login.py first")
        return False

    def _login_with_password(self, password: str) -> bool:
        try:
            self.driver.get(f"{BASE}/login")
            human_delay(2, 3)
            email_field = WebDriverWait(self.driver, 10).until(
                EC.presence_of_element_located((By.ID, "username"))
            )
            human_type(email_field, Config.CANDIDATE_EMAIL)
            human_delay(0.5, 1)
            pass_field = self.driver.find_element(By.ID, "password")
            human_type(pass_field, password)
            human_delay(0.5, 1)
            pass_field.submit()
            human_delay(4, 6)
            if self._logged_in():
                log.info("LinkedIn: password login successful")
                return True
            log.warning("LinkedIn: password login failed")
            return False
        except Exception as e:
            log.error(f"LinkedIn: password login error: {e}")
            return False

    def _logged_in(self) -> bool:
        try:
            if not self._is_alive():
                return False
            self.driver.find_element(By.CSS_SELECTOR,
                ".global-nav__me-photo, nav.global-nav, "
                ".feed-identity-module, .global-nav__primary-link")
            return True
        except NoSuchElementException:
            try:
                url = self.driver.current_url
                return any(p in url for p in ["/feed", "/mynetwork", "/jobs", "/messaging", "/notifications"])
            except Exception:
                return False

    # ── Job Search ───────────────────────────────────────────────
    def search_jobs(self, role: str, location: str) -> list[JobListing]:
        self._init()
        if not self._is_alive():
            return []
        jobs: list[JobListing] = []
        query = role.replace(" ", "%20")
        loc = location.replace(" ", "%20")

        url = (f"{BASE}/jobs/search/?keywords={query}&location={loc}"
               f"&f_E=2,3&f_AL=true&sortBy=DD&f_TPR=r604800")
        try:
            self.driver.get(url)
            human_delay(3, 5)
        except Exception:
            return []

        for _ in range(5):
            self.driver.execute_script("window.scrollBy(0, 800)")
            time.sleep(random.uniform(0.6, 1.2))

        cards = self.driver.find_elements(By.CSS_SELECTOR,
            ".jobs-search__results-list li, "
            ".scaffold-layout__list-item, "
            "li[data-occludable-job-id], div[data-job-id]")

        log.info(f"LinkedIn: {len(cards)} cards for '{role}' @ {location}")

        for card in cards[:40]:
            try:
                job = self._parse_card(card)
                if job:
                    jobs.append(job)
            except (StaleElementReferenceException, NoSuchElementException):
                continue
            except Exception:
                continue
        return jobs

    def _parse_card(self, card) -> JobListing | None:
        try:
            title_el = card.find_element(By.CSS_SELECTOR,
                "h3.base-search-card__title, a.job-card-list__title, "
                "h3[class*='title'], .job-card-container__link")
            link_el = card.find_element(By.CSS_SELECTOR,
                "a.base-card__full-link, a.job-card-list__title, a[href*='/jobs/view/']")
            try:
                company_el = card.find_element(By.CSS_SELECTOR,
                    "h4.base-search-card__subtitle, .job-card-container__company-name")
                company = company_el.text.strip()
            except NoSuchElementException:
                company = ""
            try:
                loc_el = card.find_element(By.CSS_SELECTOR,
                    ".job-search-card__location, span[class*='location']")
                location = loc_el.text.strip()
            except NoSuchElementException:
                location = ""

            url = (link_el.get_attribute("href") or "").split("?")[0]
            title = title_el.text.strip()
            if not url or not title or "/jobs/view/" not in url:
                return None
            return JobListing(platform="linkedin", title=title, company=company,
                              location=location, job_url=url)
        except (NoSuchElementException, StaleElementReferenceException):
            return None

    def _get_job_details(self, job: JobListing) -> str:
        try:
            if not self._is_alive():
                return ""
            self.driver.get(job.job_url)
            human_delay(2, 4)
            desc = WebDriverWait(self.driver, 10).until(
                EC.presence_of_element_located((By.CSS_SELECTOR,
                    ".show-more-less-html__markup, .description__text, "
                    ".jobs-description-content__text"))
            )
            return desc.text[:3000]
        except Exception:
            return ""

    # ── Easy Apply ───────────────────────────────────────────────
    def apply_to_job(self, job: JobListing) -> bool:
        if not self._is_alive():
            raise WebDriverException("LinkedIn driver session is dead")

        try:
            self.driver.get(job.job_url)
            human_delay(2, 4)
        except WebDriverException as e:
            raise

        try:
            src = self.driver.page_source.lower()
            if "you applied" in src or "application submitted" in src:
                log.info("LinkedIn: already applied", extra={"url": job.job_url})
                return False
        except Exception:
            pass

        apply_btn = None
        for sel in [
            (By.CSS_SELECTOR, ".jobs-apply-button--top-card"),
            (By.CSS_SELECTOR, "button.jobs-apply-button"),
            (By.XPATH, "//button[contains(@class,'jobs-apply-button')]"),
            (By.XPATH, "//button[contains(text(),'Easy Apply')]"),
            (By.XPATH, "//button[contains(text(),'Apply') and not(contains(text(),'Save'))]"),
        ]:
            try:
                apply_btn = WebDriverWait(self.driver, 5).until(EC.element_to_be_clickable(sel))
                break
            except (TimeoutException, NoSuchElementException):
                continue

        if not apply_btn:
            log.info("LinkedIn: no Easy Apply button", extra={"url": job.job_url})
            return False

        btn_text = apply_btn.text.strip()
        if "Apply" not in btn_text:
            return False

        try:
            safe_click(self.driver, apply_btn)
            human_delay(1, 2)
        except Exception:
            try:
                self.driver.execute_script("arguments[0].click();", apply_btn)
                human_delay(1, 2)
            except Exception:
                return False

        return self._complete_easy_apply(job)

    def _complete_easy_apply(self, job: JobListing) -> bool:
        for step in range(12):
            human_delay(1.5, 2.5)

            # Dismiss post-apply prompts
            try:
                dismiss = self.driver.find_element(By.XPATH, "//button[@aria-label='Dismiss']")
                if dismiss.is_displayed():
                    safe_click(self.driver, dismiss)
                    human_delay(0.5, 1)
                    continue
            except NoSuchElementException:
                pass

            # Fill phone number
            try:
                for phone in self.driver.find_elements(By.CSS_SELECTOR,
                        "input[id*='phoneNumber'], input[autocomplete='tel']"):
                    if phone.is_displayed() and not phone.get_attribute("value").strip():
                        human_type(phone, "9999999999")
            except Exception:
                pass

            # Fill empty text/number inputs
            try:
                for inp in self.driver.find_elements(By.CSS_SELECTOR,
                        ".jobs-easy-apply-form-element input[type='text']:not([readonly]), "
                        ".jobs-easy-apply-form-element input[type='number']:not([readonly])"):
                    if inp.is_displayed() and not inp.get_attribute("value").strip():
                        human_type(inp, "2")
                        human_delay(0.2, 0.4)
            except Exception:
                pass

            # Select radio — first per group
            try:
                seen = set()
                for radio in self.driver.find_elements(By.CSS_SELECTOR, "fieldset input[type='radio']"):
                    name = radio.get_attribute("name") or ""
                    if name not in seen and not radio.is_selected():
                        safe_click(self.driver, radio)
                        seen.add(name)
                        human_delay(0.3, 0.5)
            except Exception:
                pass

            # Handle select dropdowns
            try:
                for sel in self.driver.find_elements(By.CSS_SELECTOR,
                        ".jobs-easy-apply-form-element select"):
                    if sel.is_displayed():
                        s = Select(sel)
                        if s.first_selected_option.get_attribute("value") in ("", "Select an option"):
                            for opt in s.options:
                                if opt.get_attribute("value"):
                                    s.select_by_value(opt.get_attribute("value"))
                                    break
            except Exception:
                pass

            # Click action button
            clicked = False
            for label in ["Submit application", "Submit", "Review", "Next", "Continue"]:
                try:
                    btn = WebDriverWait(self.driver, 3).until(EC.element_to_be_clickable((By.XPATH,
                        f"//button[contains(.,'{label}') and not(contains(@class,'cancel'))]")))
                    if btn.is_displayed():
                        safe_click(self.driver, btn)
                        human_delay(1, 2)
                        if "Submit" in label:
                            self._dismiss_success_modal()
                            log.info("LinkedIn: Easy Apply submitted", extra={"title": job.title})
                            return True
                        clicked = True
                        break
                except (TimeoutException, NoSuchElementException, ElementClickInterceptedException):
                    continue

            if not clicked:
                try:
                    self.driver.find_element(By.CSS_SELECTOR, ".artdeco-modal--is-showing")
                except NoSuchElementException:
                    log.info("LinkedIn: modal closed — assumed submitted")
                    return True

        log.warning("LinkedIn: Easy Apply did not complete", extra={"url": job.job_url})
        return False

    def _dismiss_success_modal(self):
        human_delay(1, 2)
        for sel in [
            (By.XPATH, "//button[contains(@aria-label,'Dismiss')]"),
            (By.XPATH, "//button[contains(.,'Not now')]"),
            (By.XPATH, "//button[contains(.,'Done')]"),
            (By.CSS_SELECTOR, "button.artdeco-modal__dismiss"),
        ]:
            try:
                btn = WebDriverWait(self.driver, 3).until(EC.element_to_be_clickable(sel))
                safe_click(self.driver, btn)
                return
            except (TimeoutException, NoSuchElementException):
                continue

    # ── Recruiter Outreach ───────────────────────────────────────
    def send_recruiter_connection(self, recruiter_url: str) -> bool:
        if not self._is_alive():
            return False
        try:
            self.driver.get(recruiter_url)
            human_delay(2, 4)

            connect_btn = None
            for sel in [
                (By.XPATH, "//button[contains(.,'Connect')]"),
                (By.XPATH, "//button[@aria-label[contains(.,'Connect')]]"),
            ]:
                try:
                    connect_btn = WebDriverWait(self.driver, 6).until(EC.element_to_be_clickable(sel))
                    break
                except TimeoutException:
                    continue

            if not connect_btn:
                try:
                    more = WebDriverWait(self.driver, 5).until(EC.element_to_be_clickable(
                        (By.XPATH, "//button[contains(@aria-label,'More actions')]")))
                    safe_click(self.driver, more)
                    human_delay(0.5, 1)
                    connect_btn = WebDriverWait(self.driver, 5).until(EC.element_to_be_clickable(
                        (By.XPATH, "//li[contains(.,'Connect')]/div")))
                except TimeoutException:
                    return False

            safe_click(self.driver, connect_btn)
            human_delay(1, 2)

            try:
                add_note = WebDriverWait(self.driver, 5).until(EC.element_to_be_clickable(
                    (By.XPATH, "//button[contains(.,'Add a note')]")))
                safe_click(self.driver, add_note)
                note_box = WebDriverWait(self.driver, 5).until(
                    EC.presence_of_element_located((By.ID, "custom-message")))
                human_type(note_box, Config.RECRUITER_MESSAGE[:295])
                human_delay(1, 2)
            except TimeoutException:
                pass

            send_btn = WebDriverWait(self.driver, 5).until(EC.element_to_be_clickable(
                (By.XPATH, "//button[@aria-label='Send now'] | //button[contains(.,'Send')]")))
            safe_click(self.driver, send_btn)
            log.info("LinkedIn: connection request sent", extra={"url": recruiter_url})
            return True
        except Exception as e:
            log.error("LinkedIn: connection failed", extra={"error": str(e)[:100]})
            return False

    def close(self):
        if self.driver:
            try:
                self.driver.quit()
            except Exception:
                pass
            self.driver = None

File: playwright/platforms/linkedin.js
const { chromium } = require('playwright');
const path = require('path');
const fs = require('fs');
const config = require('../config');

const DELAY = (min, max) => new Promise(r => setTimeout(r, min + Math.random() * (max - min)));

function parseExpMin(expStr) {
  if (!expStr) return 0;
  const m = expStr.match(/(\d+)/);
  return m ? parseInt(m[1]) : 0;
}

class LinkedIn {
  constructor() {
    this.authFile = path.join(config.authDir, 'linkedin_state.json');
    this.browser = null;
    this.context = null;
    this.page = null;
  }

  async init() {
    const storageState = fs.existsSync(this.authFile) ? this.authFile : undefined;
    this.browser = await chromium.launch({
      headless: config.headless,
      slowMo: 40,
      args: ['--disable-blink-features=AutomationControlled'],
    });
    this.context = await this.browser.newContext({
      storageState,
      viewport: { width: 1280, height: 800 },
      userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
    });
    this.page = await this.context.newPage();
  }

  async isLoggedIn() {
    try {
      await this.page.goto('https://www.linkedin.com/feed/', { timeout: 20000, waitUntil: 'domcontentloaded' });
      await this.page.waitForSelector('.global-nav__me-photo, .feed-identity-module, nav.global-nav', { timeout: 8000 });
      return true;
    } catch {
      return false;
    }
  }

  async searchJobs(role, location) {
    const q = encodeURIComponent(role);
    const l = encodeURIComponent(location);
    const url = `https://www.linkedin.com/jobs/search/?keywords=${q}&location=${l}&f_AL=true&f_E=2,3&sortBy=DD&f_TPR=r604800`;

    try {
      await this.page.goto(url, { timeout: 30000, waitUntil: 'domcontentloaded' });
      await DELAY(2500, 4000);
    } catch {
      return [];
    }

    for (let i = 0; i < 4; i++) {
      await this.page.evaluate('window.scrollBy(0, 700)');
      await DELAY(600, 1000);
    }

    const cards = await this.page.$$('li[data-occludable-job-id], .jobs-search__results-list li, .scaffold-layout__list-item');
    const jobs = [];

    for (const card of cards.slice(0, 40)) {
      try {
        const titleEl = await card.$('a.job-card-list__title, h3.base-search-card__title, h3[class*="title"]');
        const linkEl = await card.$('a[href*="/jobs/view/"], a.base-card__full-link');
        if (!titleEl || !linkEl) continue;

        const title = (await titleEl.textContent() || '').trim();
        const href = (await linkEl.getAttribute('href') || '').split('?')[0];
        if (!href.includes('/jobs/view/') || !title) continue;

        const company = await card.$eval(
          'h4.base-search-card__subtitle, .job-card-container__company-name, a[class*="company"]',
          el => el.textContent.trim()
        ).catch(() => '');

        const location = await card.$eval(
          '.job-search-card__location, span[class*="location"]',
          el => el.textContent.trim()
        ).catch(() => '');

        const exp = await card.$eval(
          '.job-card-container__metadata-item, [class*="metadata-item"]',
          el => el.textContent.trim()
        ).catch(() => '');

        jobs.push({ title, company, location, experience: exp, url: href });
      } catch {}
    }

    return jobs;
  }

  async applyToJob(job) {
    try {
      await this.page.goto(job.url, { timeout: 30000, waitUntil: 'domcontentloaded' });
      await DELAY(2000, 3500);

      const content = await this.page.content();
      if (content.includes('You applied') || content.includes('Application submitted')) return false;

      // Find Easy Apply button
      let applyBtn = null;
      for (const sel of [
        '.jobs-apply-button--top-card',
        'button.jobs-apply-button',
        'button[class*="jobs-apply-button"]',
      ]) {
        applyBtn = await this.page.$(sel);
        if (applyBtn && await applyBtn.isVisible()) break;
        applyBtn = null;
      }
      if (!applyBtn) return false;

      const btnText = (await applyBtn.textContent() || '').trim();
      if (!btnText.includes('Apply')) return false;

      await applyBtn.click();
      await DELAY(1500, 2500);

      return await this._completeEasyApply();
    } catch {
      return false;
    }
  }

  async _completeEasyApply() {
    for (let step = 0; step < 12; step++) {
      await DELAY(1200, 2000);

      // Dismiss success modals
      try {
        const dismiss = await this.page.$('button[aria-label="Dismiss"]');
        if (dismiss && await dismiss.isVisible()) { await dismiss.click(); continue; }
      } catch {}

      // Fill phone
      try {
        for (const phone of await this.page.$$('input[id*="phoneNumber"], input[autocomplete="tel"]')) {
          if (await phone.isVisible() && !(await phone.inputValue()).trim()) {
            await phone.fill(config.phone);
          }
        }
      } catch {}

      // Fill empty text/number inputs (years of experience etc.)
      try {
        for (const inp of await this.page.$$(
          '.jobs-easy-apply-form-element input[type="text"]:not([readonly]), ' +
          '.jobs-easy-apply-form-element input[type="number"]:not([readonly])'
        )) {
          if (await inp.isVisible() && !(await inp.inputValue()).trim()) {
            await inp.fill('2');
            await DELAY(150, 300);
          }
        }
      } catch {}

      // Select first radio per group
      try {
        const seen = new Set();
        for (const radio of await this.page.$$('fieldset input[type="radio"]')) {
          const name = await radio.getAttribute('name') || '';
          if (!seen.has(name) && !(await radio.isChecked())) {
            await radio.check().catch(() => {});
            seen.add(name);
            await DELAY(200, 400);
          }
        }
      } catch {}

      // Handle dropdowns
      try {
        for (const sel of await this.page.$$('.jobs-easy-apply-form-element select')) {
          if (!await sel.isVisible()) continue;
          const val = await sel.inputValue();
          if (!val) {
            const options = await sel.$$('option[value]:not([value=""])');
            if (options.length) await sel.selectOption({ index: 1 });
          }
        }
      } catch {}

      // Click action buttons
      let clicked = false;
      for (const label of ['Submit application', 'Submit', 'Review', 'Next', 'Continue']) {
        try {
          const btn = await this.page.$(`button:has-text("${label}")`);
          if (btn && await btn.isVisible() && await btn.isEnabled()) {
            await btn.click();
            await DELAY(1200, 2000);
            if (label.includes('Submit')) {
              await this._dismissModal();
              return true;
            }
            clicked = true;
            break;
          }
        } catch {}
      }

      if (!clicked) {
        const modal = await this.page.$('.artdeco-modal--is-showing');
        if (!modal) return true;
      }
    }
    return false;
  }

  async _dismissModal() {
    await DELAY(800, 1500);
    for (const sel of [
      'button[aria-label*="Dismiss"]',
      'button:has-text("Not now")',
      'button:has-text("Done")',
      'button.artdeco-modal__dismiss',
    ]) {
      try {
        const btn = await this.page.$(sel);
        if (btn && await btn.isVisible()) { await btn.click(); return; }
      } catch {}
    }
  }

  async run(target) {
    const result = { applied: 0, failed: 0, skipped: 0, experience_counts: {}, jobs: [] };

    await this.init();

    if (!await this.isLoggedIn()) {
      console.log('[LinkedIn] Not logged in. Run: node auth_setup.js first');
      await this.browser.close();
      return result;
    }
    console.log('[LinkedIn] Logged in ✓');

    outer: for (const role of config.roles) {
      for (const location of config.locations) {
        if (result.applied >= target) break outer;

        process.stdout.write(`[LinkedIn] ${role} @ ${location} → `);
        const jobs = await this.searchJobs(role, location);
        console.log(`${jobs.length} found`);

        for (const job of jobs) {
          if (result.applied >= target) break outer;

          const exp = job.experience || 'N/A';
          result.experience_counts[exp] = (result.experience_counts[exp] || 0) + 1;

          // Skip jobs requiring more experience than user has
          if (parseExpMin(exp) > config.maxExpYears) {
            console.log(`  ↳ [${exp}] ${job.title} @ ${job.company} ... skip (over-experienced)`);
            result.skipped++;
            continue;
          }

          process.stdout.write(`  ↳ [${exp}] ${job.title} @ ${job.company} ... `);
          const ok = await this.applyToJob(job);

          if (ok) {
            result.applied++;
            result.jobs.push({ ...job, status: 'applied', platform: 'linkedin' });
            console.log('✓ Applied');
          } else {
            result.skipped++;
            console.log('skip');
          }

          await DELAY(1800, 3200);
        }
      }
    }

    // Refresh saved auth
    await this.context.storageState({ path: this.authFile });
    await this.browser.close();
    return result;
  }
}

module.exports = LinkedIn;
